ee/modules/600-flant-integration/images/madison-proxy/madison-proxy.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2021 Flant JSC
+Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https://github.com/deckhouse/deckhouse/blob/main/ee/LICENSE
+*/
+
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"net/http"
+	"net/http/httputil"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const logFormat = `request:%s %s %s, status:%s, body:%s`
+
+// This type implements the http.RoundTripper interface
+type LoggingRoundTripper struct {
+	Proxied http.RoundTripper
+}
+
+func (lrt LoggingRoundTripper) RoundTrip(req *http.Request) (res *http.Response, e error) {
+	dump, err := httputil.DumpRequestOut(req, true)
+	if err != nil {
+		log.Error(err)
+	}
+	// Send the request, get the response (or the error)
+	res, e = lrt.Proxied.RoundTrip(req)
+
+	log.Infof("%s", string(dump))
+	if res != nil {
+		log.Infof("response: %s", res.Status)
+	}
+	return
+}
+
+var (
+	listenHost  = "0.0.0.0"
+	listenPort  = "8080"
+	madisonHost = "madison.flant.com"
+)
+
+func main() {
+	madisonScheme := os.Getenv("MADISON_SCHEME")
+	if madisonScheme == "" {
+		log.Fatal("MADISON_SCHEME is not set")
+	}
+	madisonBackend := os.Getenv("MADISON_BACKEND")
+	if madisonBackend == "" {
+		log.Fatal("MADISON_BACKEND is not set")
+	}
+	madisonKey := os.Getenv("MADISON_AUTH_KEY")
+	if madisonKey == "" {
+		log.Fatal("MADISON_AUTH_KEY is not set")
+	}
+
+	proxy := newMadisonProxy(madisonScheme, madisonBackend, madisonKey)
+
+	mux := http.NewServeMux()
+
+	mux.HandleFunc("/healthz", readyHandler)
+	mux.HandleFunc("/", proxy.ServeHTTP)
+
+	s := &http.Server{
+		Addr:    listenHost + ":" + listenPort,
+		Handler: mux,
+	}
+	go func() {
+		err := s.ListenAndServe()
```

Comment:
How would the main goroutine exit if there was an error in this goroutine? I assume that the kubernetes liveness probe will help us there, yet it is not always right to rely on external systems behavior.

Reply:
added second channel